Added initial plant critter behavior

treivize opened the pull request:
Forum thread: [Plant Critter](https://forum.unity.com/threads/plant-critter.1005700/)
Codeck card: [Plant Critter Integration](https://open.codecks.io/unity-open-project-1/decks/15-code/card/186-animation-integration)

![image](https://user-images.githubusercontent.com/42570903/100622851-a92a6f80-3321-11eb-98e6-6fdabf294122.png)

DeivSky commented on UOP1_Project/Assets/Scripts/Characters/StateMachine/Actions/FaceProtagonistSO.cs:
```diff
@@ -0,0 +1,37 @@
+	
+	public override void OnStateEnter()
+	{
+		_protagonist = GameObject.FindObjectOfType<Protagonist>().transform;
```
You should cache this on `Awake`, or, actually, it should be in a field in `FaceProtagonistSO` so it can be set from the inspector. Also, I'd rename the action to `FaceTarget` since this could be useful for other things as well.

treivize replied:
Actually, I initially put this in Awake, but it does not work because here the protagonist is instantiated by the spawning system after Awake is triggered. 

I was thinking also about having a target more generic not Protagonist specific, but how do you specify a target that has not been instantiated yet in the SO?

DeivSky replied:
> Actually, I initially put this in Awake, but it does not work because here the protagonist is instantiated by the spawning system after Awake is triggered.
Hmm, I've actually had this at the back of my mind for a while... The **SM** is calling `Awake` on the initial state _during_ the **SM** `Awake`. I think it should be called on `Start` to solve this.

ciro-unity replied:
@treivize that's a good question.

Actually, one of the last things I implemented might be an answer to this. [I mentioned it here](https://forum.unity.com/threads/decouple-cameramanager-and-spawnsystem-card-194.1016932/#post-6592348).

You could use one of those to connect this ActionSO to the player's transform, even if you don't know yet if it's been instantiated or not. You could just implement a check to see if it's null. If it's not, it will be the player's transform.

The check is also not super necessary, since it will only happen if the player has entered the TriggerZone, so the player exists at that point.

DeivSky commented on UOP1_Project/Assets/Scripts/Characters/StateMachine/Actions/FaceProtagonistSO.cs:
```diff
@@ -0,0 +1,37 @@
+	{
+		if (_protagonist != null)
+		{
+			_actor.LookAt(_protagonist);
```
Note that this is creating some probably undesired behavior!
![image](https://user-images.githubusercontent.com/27371812/100672241-67f58800-3340-11eb-9c96-6a474485e765.png)
😆 
Rotation should be only on the Y axis.

DeivSky commented on UOP1_Project/Assets/Scripts/Characters/EnemyColliderZoneController.cs:
```diff
@@ -0,0 +1,26 @@
+
+	[SerializeField] private VoidEventChannelSO _exitZone = default;
+
+	private void OnTriggerEnter(Collider other)
+	{
+		if (other.GetComponent<Protagonist>() != null)
+		{
+			_enterZone.RaiseEvent();
+		}
+	}
+
+	private void OnTriggerExit(Collider other)
+	{
+		if (other.GetComponent<Protagonist>() != null)
+		{
+			_exitZone.RaiseEvent();
+		}
+	}
```
This could be a very generic `ZoneTriggerController` if you detected the collision by layer:
``` csharp
[SerializeField] private LayerMask _layers = default;

private void OnTriggerEnter(Collider other)
{
	if (((1 << other.gameObject.layer) & _layers) != 0)
	{
		// stuff
	}
}
```

DeivSky reviewed (changes requested):
Hey, looking great! I'm leaving just a few comments.

And besides that, I have to ask, are you enjoying the StateMachine? Feel free to leave any feedback in the [thread](https://forum.unity.com/threads/generic-state-machine.979371/)! 😃

DeivSky commented on UOP1_Project/Assets/Scripts/StateMachine/Editor/Utilities/TransitionDisplayHelper.cs:
```diff
@@ -46,7 +46,14 @@ internal bool Display(ref Rect position)
 
 				rect.x += 20;
-				LabelField(rect, SerializedTransition.ToState.objectReferenceValue.name, EditorStyles.boldLabel);
+				if (SerializedTransition.ToState.objectReferenceValue != null)
+				{
+					LabelField(rect, SerializedTransition.ToState.objectReferenceValue.name, EditorStyles.boldLabel);
+				}
+				else
+				{
+						LabelField(rect, "Unknown", EditorStyles.boldLabel);
+				}
```
Actually, I am checking and logging it in `TransitionTableEditor.cs`, but there's an error in the script!
``` csharp
if (serializedTransition.FromState.objectReferenceValue == null)
{
	Debug.LogError("Transition with invalid state found in table " + serializedObject.targetObject.name + ", deleting...");
	_transitions.DeleteArrayElementAtIndex(i);
	serializedObject.ApplyModifiedProperties();
	Reset();
	return;
}
if (serializedTransition.FromState.objectReferenceValue == null)
{
	Debug.LogError("Transition with invalid state found in table " + serializedObject.targetObject.name + ", deleting...");
	_transitions.DeleteArrayElementAtIndex(i);
	serializedObject.ApplyModifiedProperties();
	Reset();
	return;
}
```
I'm checking `FromState` twice!! 🤦🏻

treivize replied:
My bad, indeed, I should not have added this null check in the middle of this PR... 
Indeed, I did that, because I experienced fatal exception in case of state deletion: Transition Table was crashing when opening it if a ToState was missing! 

I better understand why with your last comment! It is indeed better like that :) I will update my PR to remove the null check!

BTW: I did not follow actively the SM thread (I am already bothered on other topics and I am mostly using my free time to contribute), despite that it was easy to understand how it works and how to use it to build the behavior I wanted to put in place, especially without having to write lots of boiler plate code! Great job! 
BTW2: I have also learned lots of things reading the code you have written here!

ciro-unity commented:
Hey, I merged the fix by @DeivSky on #244. Does that affect you in any way?

treivize commented:
> Hey, I merged the fix by @DeivSky on #244. Does that affect you in any way?

No, there is impact, the issue was when building the transition table when deleting or renaming states, but here everything is wired properly. And in my last commit, I have removed the null check I had introduced initially in TransitionDisplayHelper, but was not needed after the fix from DeivSky.

It is ok to merge this one, the implementation is simple and most of the things are right. But as described in my last post in Plant Critter thread, I have another version in the pipe, that might be more future proof. I have the code in another branch deriving from this one:

[Fork PR](https://github.com/treivize/open-project-1/pull/1)

ciro-unity commented:
Honestly I think this one is better, the idea of using names and strings for finding the player is pretty dangerous.

I've also written it in one of the comment (on this PR), but regarding your hard time in how to find the player's transform, we might use the RuntimeAnchors concept I've implemented in the project as of late. [I explained it here](https://forum.unity.com/threads/decouple-cameramanager-and-spawnsystem-card-194.1016932/#post-6592348).

So if you want, I can go ahead and merge this, implementing the anchor in it.

treivize commented:
> Honestly I think this one is better, the idea of using names and strings for finding the player is pretty dangerous.
> 
> I've also written it in one of the comment (on this PR), but regarding your hard time in how to find the player's transform, we might use the RuntimeAnchors concept I've implemented in the project as of late. [I explained it here](https://forum.unity.com/threads/decouple-cameramanager-and-spawnsystem-card-194.1016932/#post-6592348).
> 
> So if you want, I can go ahead and merge this, implementing the anchor in it.

Definitely this one looks more robust, the only aspect that I do not like this fact that there is no link between the element entering into the alert zone and what the plant is facing which is always the Protagonist, assuming that it is necessary him who has entered the zone...  Maybe I should not care since it might be the case in the type of game we are building. 

Just regarding the other version, there is no hard coded prefab name expected in the SO, the name is fetched from the collider detection script and injected into the SO.

For the change to be made in the FaceProtagonistSO, I think I can make it, I was wondering if it could make more sense to retrofit my V2 approach using the new TransformEventChannelSO to send the ref of the element entering the collider to the facing script?

ciro-unity commented:
> For the change to be made in the FaceProtagonistSO, I think I can make it, I was wondering if it could make more sense to retrofit my V2 approach using the new TransformEventChannelSO to send the ref of the element entering the collider to the facing script?

I think it's a bit more safe if you go with the RuntimeAnchor stuff I was talking about. I wouldn't necessarily trust two actions in the StateMachine to send events to each other, if we change anything in the SM (i.e. introduce some kind of caching, recycling, etc.) that might introduce a bug.
With the Anchor way, if the protagonist is in the scene at all, that Transform reference will be there. And if for some reason it's not, there's a lot of other systems that will fail too giving us a lot of warning.

treivize commented:
No problem, I understand. I am fine with it. I am doing the change to use PlayerTransform SO in FaceProtagonist script, I thought that this SO was already set somewhere, but it seems it is not the case, am I wrong? I am planning to set it in SpawnSystem right after the player instantiation.

ciro-unity commented:
> I thought that this SO was already set somewhere, but it seems it is not the case, am I wrong?

Sorry, I should have really just linked it (since it's in the repo). The RuntimeAnchor is in this folder: https://github.com/UnityTechnologies/open-project-1/tree/main/UOP1_Project/Assets/ScriptableObjects/RuntimeAnchors

treivize commented:
> > I thought that this SO was already set somewhere, but it seems it is not the case, am I wrong?
> 
> Sorry, I should have really just linked it (since it's in the repo). The RuntimeAnchor is in this folder: https://github.com/UnityTechnologies/open-project-1/tree/main/UOP1_Project/Assets/ScriptableObjects/RuntimeAnchors

Actually I had found it, but what I meant is I did not find where PlayerTransform is initialized with the Player transform. So when I used it in my FaceProtagonist, the inner check isSet was always false. That is why I have added the PlayerTransfrom anchor to the SpawnSystem to be set right after its instantiation. 
![image](https://user-images.githubusercontent.com/42570903/101330015-69f4a500-3872-11eb-9829-4a8bfffd619b.png)

ciro-unity commented:
Yes, I was confused for a second because I thought that I had done it already, but no... I just prepared the setup :D
So yeah, that's where you'd do it. Also because the spawn will change a bit (to support multiple entry points) so it's that which should take care of spawning and destroying the player prefab.

treivize commented:
Ok, I have just moved this logic in Spawn() method next to _playerInstantiatedChannel event raising, I think it makes more sense to be here instead of the gameObject instantiation from prefab method (InstantiatePlayer()).

I am not planning to push any further update to this PR, it should be fine now.

DapperDino commented:
What about if you have multiple instances of the same enemy? Won't the scriptable object events become an issue? If one enemy is alerted of the player and changes state, won't all enemies receive that event and switch state? Maybe I'm misunderstanding the use of events here?

treivize commented:
> What about if you have multiple instances of the same enemy? Won't the scriptable object events become an issue? If one enemy is alerted of the player and changes state, won't all enemies receive that event and switch state? Maybe I'm misunderstanding the use of events here?

oh god! you are perfectly right! Indeed the event is shared across all instances :( So they will act in sync what ever where they are from the player...

DapperDino commented:
Do you have any solutions for this that still uses Scriptable Objects or will we have to revert to using normal C# events?

treivize commented:
> Do you have any solutions for this that still uses Scriptable Objects or will we have to revert to using normal C# events?

I have no smart ideas right now... I guess I have to go using UnityEvent editable in the inspector and wiring the actions... Even for the conditions I am not sure how to implement it they are based on SO, so I have the feeling that using UnityEvent to trigger a function in it will do it for all the instances of the plant. I have to test it

DapperDino commented:
Let me know how it goes and I'll try coming up with a solution in the meantime 👍

treivize commented:
Ok! Here is my proposition to handle this problem:

I am storing the collider state into a new MonoBehavior script "Critter":
![image](https://user-images.githubusercontent.com/42570903/101398563-93431e80-38ce-11eb-84e1-68b5d3268abb.png)

The data are set by each ZoneTriggerController associated to the collider via UnityEvent<bool>
![image](https://user-images.githubusercontent.com/42570903/101399012-3300ac80-38cf-11eb-9aef-11b215fa5478.png)

Then the data are read by a revamp SM ConditionSO script:
![image](https://user-images.githubusercontent.com/42570903/101399149-5a577980-38cf-11eb-9ca1-acf1597af992.png)

treivize commented:
Should this PR target main or art-assets?

treivize commented:
Just an updated version of the implementation diagram:
![image](https://user-images.githubusercontent.com/42570903/102541726-0dcf2380-40b1-11eb-8ad6-eeae9ca75546.png)

ciro-unity commented:
Great PR, @treivize! I'm going to merge it soon, maybe with a few minor adjustments and also I want to get the other one you made and merge them together.

What is the big cylinder collider doing on the base object though?
![image](https://user-images.githubusercontent.com/20049224/104120964-2abae780-533b-11eb-858c-6ef11a0100c8.png)

ciro-unity commented:
Ah, I guess it's to impede the character from jumping over it? Never thought about this. Do we want to stop that?

ciro-unity commented:
Also another question: why all the getter/setter on the variables in the `Critter` script? `_isPlayerInAttackZone`, `_playerInAlertZone`... They are not serving a purpose right now, why not just make the variables public?

treivize commented:
Ah! So happy they fit with what is expected by the project!

Indeed it was made to prevent the player to jump on its head. Maybe it would be preferable to prevent it differently by repulsing the player if attempted. Just the plant critter attack will be a bit odd (not facing the player) if the player is standing up on its head. 

In the other PR I was thinking about disabling some exit time in the animator of the pig and the critter to make the attack and the hit more aligned.

treivize commented:
About the getter/setter no particular reason. I was looking at some guidelines in C# and how to expose data from a simple class and I found this pattern. Honestly, I feel the same, public is equivalent and less verbose.
I will push an update tonight accordingly.

ciro-unity commented:
No, don't worry about modifying this PR, I'm merging some stuff so I can do it.

ciro-unity commented:
I actually realised one use for the getter/setters: you're using them in the UnityEvent on the Critter's ZoneColliders, to feed the boolean value directly as part of the event being passed along. So I guess they can stay, maybe I'll add a comment in the script to explain it.

treivize commented:
Indeed but I think Unity event should be able to get/set public properties as well, right?

Are you planning to merge both PR together? Because I had in mind the following improvements for the combat part :
1. Improve the ActiveWeaponActionSO to activate the weapon only during a sub part of a state with a delay from start plus duration, since the attack animation contains a start and an end that should not be armful.
2. Change the Killed trigger by a Boolean in plant critter animator 
3. Add a "Fatal Hit" state followed immediately by the faint animation in the plant critter with a transition when killed equals true. It should replace the current faint state.
4. Add killed == false in the transition to GetHit in the plant critter animator and unchecked has exit time for all transition to get hit and fatal hit 
5. Unchecked has exit time in pig chef animator for the transition between walking and canehit 

Just in case you would be interested in including some of them in your merge otherwise it can be submitted in a later PR.

ciro-unity commented:
> Indeed but I think Unity event should be able to get/set public properties as well, right?

Yes, but apparently you still need to provide getters/setters... TIL 🤷🏻‍♂️ 
So I will just shorten them to the classic `{ get; set; }` syntax.



> Are you planning to merge both PR together? Because I had in mind the following improvements for the combat part :

Yes, I was going to merge it all before you make more adjustments, if you don't mind. I can tell you I had already implemented 2 and 5. For the rest, are you thinking of a random critical hit?

treivize commented:
Not a critical hit, even if it might be a good idea 👍 
I think the current animator for the plant critter is not optimal when fainting, because, currently the last hit before dying sometimes do not trigger the get hit animation, but directly the fainting one from alert, which looks a bit odd to me.
I will have a look and try to propose an enhancement from your merge.

BTW, I am very glad to see these PRs finally integrated in the project!

treivize commented:
I just pull the project after your merge and release that you actually implemented the change I just had in mind for the plant critter animator and the faint transition (making transitioning from BeenHit state only) I did some tests and I do not know yet why, but sometimes the plant is dying without going through the expected animator states, just stays in Alert and disappear after a while!
I am trying to understand why...
